src/domain/core/src/services/compact_service.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::sync::Arc;
+
+use opendatafabric::*;
+use thiserror::Error;
+
+use crate::*;
+
+#[async_trait::async_trait]
+pub trait CompactService: Send + Sync {
+    async fn compact_dataset(
+        &self,
+        dataset_handle: &DatasetHandle,
+        max_slice_size: u64,
+        max_slice_records: u64,
+        listener: Option<Arc<dyn CompactionMultiListener>>,
+    ) -> Result<(), CompactError>;
```

Comment:
Changed

Reply:
```rust
pub enum CompactResult {
    Finished,
    CleanUpRequired,
}
```
This is not a very informative result. It's more confusing than anything - does `CleanUpRequired` mean that the compaction didn't finish? I don't see `CleanUpRequired` returned anywhere.

I'd like to get at least a clear indication in response whether compaction:
- was performed (including info like old_head, new_head, old_num_blocks, new_num_blocks)
- was not required - head remains the same

Reply:
I was thinking that we have two important phases(first one all stages before cleaning up and after). If we get any error during the first phase we will return the error, which says us compaction wasn't successful and we did not change the provided dataset. But if we provide the stage of cleaning old files and we get here an error we should not return an error, because we already compacted the dataset but we will notify that cleaning is still required.

But, yeah it is nice to have such params.
Added